File: api-fiber/internal/handlers/todo.go
package handlers

import (
	"context"
	"log"
	"time"

	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"
	"todo-app/internal/db"
	apperrors "todo-app/internal/errors"
	"todo-app/internal/models"
)

// ListTodos retrieves all todos from the database
func ListTodos(c *fiber.Ctx) error {
	query := `
		SELECT id, title, description, completed, created_at, updated_at
		FROM todos
		ORDER BY created_at DESC
	`

	rows, err := db.Pool.Query(context.Background(), query)
	if err != nil {
		log.Printf("Error querying todos: %v\n", err)
		return apperrors.HandleError(c, apperrors.NewInternalServerError("Failed to fetch todos"))
	}
	defer rows.Close()

	var todos []models.TodoResponse
	for rows.Next() {
		var todo models.Todo
		err := rows.Scan(&todo.ID, &todo.Title, &todo.Description, &todo.Completed, &todo.CreatedAt, &todo.UpdatedAt)
		if err != nil {
			log.Printf("Error scanning todo: %v\n", err)
			return apperrors.HandleError(c, apperrors.NewInternalServerError("Failed to process todos"))
		}

		var response models.TodoResponse
		response.FromTodo(&todo)
		todos = append(todos, response)
	}

	if err = rows.Err(); err != nil {
		log.Printf("Error iterating todos: %v\n", err)
		return apperrors.HandleError(c, apperrors.NewInternalServerError("Failed to fetch todos"))
	}

	if todos == nil {
		todos = []models.TodoResponse{}
	}

	return c.Status(fiber.StatusOK).JSON(todos)
}

// GetTodo retrieves a single todo by ID
func GetTodo(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return apperrors.HandleError(c, apperrors.NewBadRequest("Invalid todo ID format"))
	}

	query := `
		SELECT id, title, description, completed, created_at, updated_at
		FROM todos
		WHERE id = $1
	`

	var todo models.Todo
	err = db.Pool.QueryRow(context.Background(), query, id).Scan(
		&todo.ID, &todo.Title, &todo.Description, &todo.Completed, &todo.CreatedAt, &todo.UpdatedAt,
	)
	if err != nil {
		log.Printf("Error querying todo: %v\n", err)
		return apperrors.HandleError(c, apperrors.NewNotFound("Todo not found"))
	}

	var response models.TodoResponse
	response.FromTodo(&todo)

	return c.Status(fiber.StatusOK).JSON(response)
}

// CreateTodo creates a new todo
func CreateTodo(c *fiber.Ctx) error {
	var req models.CreateTodoRequest
	if err := c.BodyParser(&req); err != nil {
		return apperrors.HandleError(c, apperrors.NewBadRequest("Invalid request body"))
	}

	// Validate title is not empty
	if req.Title == "" {
		return apperrors.HandleError(c, apperrors.NewBadRequest("Title is required and cannot be empty"))
	}

	id := uuid.New()
	now := time.Now().UTC()

	query := `
		INSERT INTO todos (id, title, description, completed, created_at, updated_at)
		VALUES ($1, $2, $3, $4, $5, $6)
		RETURNING id, title, description, completed, created_at, updated_at
	`

	var todo models.Todo
	err := db.Pool.QueryRow(
		context.Background(),
		query,
		id, req.Title, req.Description, false, now, now,
	).Scan(&todo.ID, &todo.Title, &todo.Description, &todo.Completed, &todo.CreatedAt, &todo.UpdatedAt)

	if err != nil {
		log.Printf("Error creating todo: %v\n", err)
		return apperrors.HandleError(c, apperrors.NewInternalServerError("Failed to create todo"))
	}

	var response models.TodoResponse
	response.FromTodo(&todo)

	return c.Status(fiber.StatusCreated).JSON(response)
}

// UpdateTodo updates an existing todo
func UpdateTodo(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return apperrors.HandleError(c, apperrors.NewBadRequest("Invalid todo ID format"))
	}

	var req models.UpdateTodoRequest
	if err := c.BodyParser(&req); err != nil {
		return apperrors.HandleError(c, apperrors.NewBadRequest("Invalid request body"))
	}

	// Check if todo exists
	checkQuery := `SELECT id FROM todos WHERE id = $1`
	var existingID uuid.UUID
	err = db.Pool.QueryRow(context.Background(), checkQuery, id).Scan(&existingID)
	if err != nil {
		log.Printf("Error checking todo existence: %v\n", err)
		return apperrors.HandleError(c, apperrors.NewNotFound("Todo not found"))
	}

	// Build dynamic update query
	updateQuery := `
		UPDATE todos
		SET
			title = COALESCE($2, title),
			description = COALESCE($3, description),
			completed = COALESCE($4, completed),
			updated_at = $5
		WHERE id = $1
		RETURNING id, title, description, completed, created_at, updated_at
	`

	var todo models.Todo
	err = db.Pool.QueryRow(
		context.Background(),
		updateQuery,
		id,
		req.Title,
		req.Description,
		req.Completed,
		time.Now().UTC(),
	).Scan(&todo.ID, &todo.Title, &todo.Description, &todo.Completed, &todo.CreatedAt, &todo.UpdatedAt)

	if err != nil {
		log.Printf("Error updating todo: %v\n", err)
		return apperrors.HandleError(c, apperrors.NewInternalServerError("Failed to update todo"))
	}

	var response models.TodoResponse
	response.FromTodo(&todo)

	return c.Status(fiber.StatusOK).JSON(response)
}

// DeleteTodo deletes a todo by ID
func DeleteTodo(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return apperrors.HandleError(c, apperrors.NewBadRequest("Invalid todo ID format"))
	}

	// Check if todo exists
	checkQuery := `SELECT id FROM todos WHERE id = $1`
	var existingID uuid.UUID
	err = db.Pool.QueryRow(context.Background(), checkQuery, id).Scan(&existingID)
	if err != nil {
		log.Printf("Error checking todo existence: %v\n", err)
		return apperrors.HandleError(c, apperrors.NewNotFound("Todo not found"))
	}

	deleteQuery := `DELETE FROM todos WHERE id = $1`
	_, err = db.Pool.Exec(context.Background(), deleteQuery, id)
	if err != nil {
		log.Printf("Error deleting todo: %v\n", err)
		return apperrors.HandleError(c, apperrors.NewInternalServerError("Failed to delete todo"))
	}

	return c.SendStatus(fiber.StatusNoContent)
}

File: api-fastify/src/handlers/todo.js
import { v4 as uuidv4 } from 'uuid';
import { query } from '../db/connection.js';
import { Todo } from '../models/todo.js';
import logger from '../utils/logger.js';

// List all todos
export const listTodos = async (request, reply) => {
  try {
    const result = await query(
      'SELECT id, title, description, completed, created_at, updated_at FROM todos ORDER BY created_at DESC'
    );

    const todos = result.rows.map(
      (row) =>
        new Todo(
          row.id,
          row.title,
          row.description,
          row.completed,
          row.created_at,
          row.updated_at
        )
    );

    return reply.send(todos);
  } catch (err) {
    logger.error(err, 'Error listing todos');
    return reply.status(500).send({
      error: 'INTERNAL_SERVER_ERROR',
      message: 'Failed to fetch todos',
    });
  }
};

// Get a specific todo
export const getTodo = async (request, reply) => {
  try {
    const { id } = request.params;

    // Validate UUID format
    if (!isValidUUID(id)) {
      return reply.status(400).send({
        error: 'BAD_REQUEST',
        message: 'Invalid todo ID format',
      });
    }

    const result = await query(
      'SELECT id, title, description, completed, created_at, updated_at FROM todos WHERE id = $1',
      [id]
    );

    if (result.rows.length === 0) {
      return reply.status(404).send({
        error: 'NOT_FOUND',
        message: 'Todo not found',
      });
    }

    const row = result.rows[0];
    const todo = new Todo(
      row.id,
      row.title,
      row.description,
      row.completed,
      row.created_at,
      row.updated_at
    );

    return reply.send(todo);
  } catch (err) {
    logger.error(err, 'Error getting todo');
    return reply.status(500).send({
      error: 'INTERNAL_SERVER_ERROR',
      message: 'Failed to fetch todo',
    });
  }
};

// Create a new todo
export const createTodo = async (request, reply) => {
  try {
    const { title, description } = request.body;

    // Validate title
    if (!title || title.trim() === '') {
      return reply.status(400).send({
        error: 'BAD_REQUEST',
        message: 'Title is required and cannot be empty',
      });
    }

    const id = uuidv4();
    const now = new Date();

    const result = await query(
      `INSERT INTO todos (id, title, description, completed, created_at, updated_at)
       VALUES ($1, $2, $3, $4, $5, $6)
       RETURNING id, title, description, completed, created_at, updated_at`,
      [id, title, description || null, false, now, now]
    );

    const row = result.rows[0];
    const todo = new Todo(
      row.id,
      row.title,
      row.description,
      row.completed,
      row.created_at,
      row.updated_at
    );

    return reply.status(201).send(todo);
  } catch (err) {
    logger.error(err, 'Error creating todo');
    return reply.status(500).send({
      error: 'INTERNAL_SERVER_ERROR',
      message: 'Failed to create todo',
    });
  }
};

// Update a todo
export const updateTodo = async (request, reply) => {
  try {
    const { id } = request.params;
    const { title, description, completed } = request.body;

    // Validate UUID format
    if (!isValidUUID(id)) {
      return reply.status(400).send({
        error: 'BAD_REQUEST',
        message: 'Invalid todo ID format',
      });
    }

    // Check if todo exists
    const existsResult = await query('SELECT id FROM todos WHERE id = $1', [id]);
    if (existsResult.rows.length === 0) {
      return reply.status(404).send({
        error: 'NOT_FOUND',
        message: 'Todo not found',
      });
    }

    const now = new Date();

    // Build dynamic update query
    const updates = [];
    const params = [id];
    let paramCount = 2;

    if (title !== undefined && title !== null) {
      updates.push(`title = $${paramCount}`);
      params.push(title);
      paramCount++;
    }

    if (description !== undefined) {
      updates.push(`description = $${paramCount}`);
      params.push(description);
      paramCount++;
    }

    if (completed !== undefined && completed !== null) {
      updates.push(`completed = $${paramCount}`);
      params.push(completed);
      paramCount++;
    }

    updates.push(`updated_at = $${paramCount}`);
    params.push(now);

    const updateQuery = `
      UPDATE todos
      SET ${updates.join(', ')}
      WHERE id = $1
      RETURNING id, title, description, completed, created_at, updated_at
    `;

    const result = await query(updateQuery, params);

    const row = result.rows[0];
    const todo = new Todo(
      row.id,
      row.title,
      row.description,
      row.completed,
      row.created_at,
      row.updated_at
    );

    return reply.send(todo);
  } catch (err) {
    logger.error(err, 'Error updating todo');
    return reply.status(500).send({
      error: 'INTERNAL_SERVER_ERROR',
      message: 'Failed to update todo',
    });
  }
};

// Delete a todo
export const deleteTodo = async (request, reply) => {
  try {
    const { id } = request.params;

    // Validate UUID format
    if (!isValidUUID(id)) {
      return reply.status(400).send({
        error: 'BAD_REQUEST',
        message: 'Invalid todo ID format',
      });
    }

    // Check if todo exists
    const existsResult = await query('SELECT id FROM todos WHERE id = $1', [id]);
    if (existsResult.rows.length === 0) {
      return reply.status(404).send({
        error: 'NOT_FOUND',
        message: 'Todo not found',
      });
    }

    await query('DELETE FROM todos WHERE id = $1', [id]);

    return reply.status(204).send();
  } catch (err) {
    logger.error(err, 'Error deleting todo');
    return reply.status(500).send({
      error: 'INTERNAL_SERVER_ERROR',
      message: 'Failed to delete todo',
    });
  }
};

// Helper function to validate UUID
function isValidUUID(uuid) {
  const uuidRegex =
    /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
  return uuidRegex.test(uuid);
}
